doc/developer/design/20240609_error_handling.md

```diff
@@ -0,0 +1,440 @@
+nor the source *errs stream*.
+
+When `INLINE` is specified, the source's relation will contain two
+top-level columns, `error` and `value` and will not contain top-level columns for
+fields in decoded values. The column named `error` is nullable with a
+type of `record(description: text, code: text)`. If the most recent value for a
+key has been successfully decoded, this column will be `NULL`. If the most recent
+value for a key was not succesfully decoded, this column will contain details
+about the error. In this case the value is forced into the nullable column
+named `value` with a type reflecting the configured value. Format—flattening of
+record-type values into top-level columns does not occur.
+
+When both `INLINE` and `PROPAGATE` are specified, the errors are both reported
+inline in the source and propagated to the DLQ table or the source's *errs
+stream*, whichever is active.
+
+Even when using `VALUE DECODING ERRORS = INLINE`, users need to monitor the DLQ
+table or the source's *errs stream* for errors, as errors while decoding the key
+still get sent to the DLQ. (There is no good way to represent key decoding
+errors inline without breaking upsert semantics.)
+
+The `INLINE` behavior allows users to discover upsert sources that presently
+have bad data by querying each source and checking `count(error) > 0`. They can
+discover the affected keys by running `SELECT key ... WHERE error IS NOT NULL`.
+
+**Open questions:**
+
+  * Does the separation of the `REDIRECT ERRORS` and the `VALUE DECODING ERRORS`
+    options make sense?
+    - 6/18 Update: We are considering just implementing the `VALUE DECODING ERRORS`
+    approach to start, as this provides correctness guarantees and solves the immediate
+    need of our customers who primarily encounter value-decoding errors in Kafka
+    upsert sources.
+
+#### Limitations
+
+The above options do not generalize to handling decoding errors that occur
+outside of the source decoding pipeline. For example, imagine a JSON source that
+is further parsed into columns by a downstream materialized view:
+
+```sql
+CREATE SOURCE my_source
+  FROM KAFKA CONNECTION kafka_connection (TOPIC 'samsa')
+  FORMAT JSON;
+
+CREATE MATERIALIZED VIEW my_view AS
+  SELECT
+    (data->>'userid')::int AS userid,
+  FROM my_source;
+```
+
+Bad data can cause both the initial JSON parsing to fail, or the cast to `int`
+in `my_view` to fail. The `REDIRECT ERRORS` and `VALUE DECODING ERRORS` options
+do not help with the second type of error. We'll likely need to additionally
+pursue a solution for gracefully handling invalid function calls, like those
+described in [#6367].
+
+### Implementation
+
+TODO.
+
+**Open questions:**
+
+  * How do we ensure consistency? The DLQ shard and data shard need to be
+    updated in lockstep.
+
+  * When using `VALUE DECODING ERRORS = INLINE`, do we correctly handle
+    retractions across versions if the error text changes? (How do we handle
+    this today?)
```

Comment:
I'm slightly confused by this question -- can anyone clarify if my understanding below is incorrect:

- An upsert source receives a (key, value) pair and checks its backend (e.g. rocksdb) to see if the key is already present. If it is already present, it emits a diff `((key, old_value), -1)` to retract the old value for the key before emitting `((key, value), 1)` to signify the new value.
- When an upsert source is restarted, it reads back its own collection from persist, which may or may not have compacted away some of the previous data whose diffs cancel out.
- This is used to rehydrate the upsert backend (rocksdb) with the present value of each key

Let's say a key contained an inline error-value with some error text at version foo: `(key, error-value-foo)` and this was rehydrated into a new instance running code at version bar.

The error value read from persist on rehydration would still be `(key, error-value-foo)`. And if the bad value was re-read from the upstream source such that the decoding error generated `(key, error-value-bar)`, we would just emit a `((key, error-value-foo), -1)`, `((key, error-value-bar), 1)`, right?

In what case might we accidentally miss a retraction if the error text changed between versions?